Blog post on video player MSE API (#471)
BedrockStreaming/tech.bedrockstreaming.com: _posts/2025-06-23-video-player-mse-introduction.md

---
layout: post
title: "Crafting Your Own Streamer: A Hands-on Guide to JavaScript MSE"
description: In this article, I'm going to explain how to enhance your player using the Media Source Extension API.
author: [a_gaudard]
tags: [TVJS, smartTV, web, javascript, frontend, player, video]
color: rgb(251,87,66)
language: en
---

Streaming is now everywhere and used by almost everyone, but do you know how it works behind the frame?

> Let say you want to create a player to watch your favorite TV Show.

Reviewer: ```suggestion
Let's say you want to create a player to watch your favorite TV Show.
```

First step will probably be to create a page where you will add the famous video tag, give it a source and maybe display the controls so you can interact with it.
You will end up with something like this:

```html
<video src="my-awesome-video" controls></video>
```

> Your video will probably play well, but depending on, for example the video size or your connexion, you might have some performances issues.

Reviewer: ```suggestion
Your video will probably play well, but depending on, for example, the video size or your connection, you might have some performances issues.
```

Here MDN tells us that MSE allows us to better control the flow of the stream.

**This means new features are now available!**

Going back to my previous use case, what if I want to play Dune in 4K?
It is going to be a super heavy file and if I give it to my video tag, it might download the entire file before being ready to play the content.

Do I want to wait to watch my movie? Of course not! I'm already on the sofa with my ice cream ready for it to start 🍦

MSE helps us manipulate specific video file formats such as Dash (Dynamic Adaptive Streaming over HTTP) or HLS (HTTP Live Streaming). Those formats split a video into lists of chunks and allow playing a short section of the video at any timecode. Those chunks are referenced in a manifest, .mpd for DASH and .m3u8 for HLS.

![Manifest file pointing to segments](/images/posts/2025-06-23-video-player-mse-introduction/manifest-1.png)

That's not all! Those manifest can transport additional informations.

We can now have list of chunks based on specific settings such as the resolution, a recommanded bandwidth per list, a link to subtitles and many more!

> Having multiple playlist based on the resoltion brings a new acronym (I now what you think another one). We can now talk about Adaptive Bitrate Streaming (= ABR).

Reviewer: ```suggestion
Having multiple playlist based on the resoltion brings a new acronym (I know what you think, another one). We can now talk about Adaptive Bitrate Streaming (= ABR).
```

### Step 2

Now we want to add a **MediaSource** to handle our manifest.

Also, to better manipulate our video tag we use the useRef hook from React.

```js
import { useEffect, useRef } from "react";

export const Player = ({ source }: { source: string }) => {
  const playerRef = useRef<HTMLVideoElement>(null);

  // MediaSource instance
  let ms: MediaSource;

  useEffect(() => {
    ms = new MediaSource();

    // Link the MediaSource to the video tag
    playerRef.current!.src = window.URL.createObjectURL(ms);

    // 1 - Handle the sourceopen event
    ms.addEventListener('sourceopen', onMediaSourceOpen);
  });

  return (<>
    <video title="Video player" controls ref={playerRef} />
  </>)
}
```

### Step 3

Now we need to add a SourceBuffer to handle each segments of our video.

```js
import { useEffect, useRef } from "react";

export const Player = ({ source }: { source: string }) => {
  const playerRef = useRef<HTMLVideoElement>(null);

  let ms: MediaSource;

  // SourceBuffer instance
  let videoSourceBuffer: SourceBuffer;

  // 1 - Define the SourceBuffer
  const onMediaSourceOpen = () => {
    videoSourceBuffer = ms.addSourceBuffer('video/mp4; codecs="avc1.42c01e"');

>     // 2 - We add a new video chunk on each SourceBuffer update

Reviewer: nit:
```suggestion
    // 2 - We fetch the next video chunk after each SourceBuffer update
```

  const nextVideoSegment = () => {
    var url = segmentsVideo.replace('$Number$', indexVideoSegment.toString());

    fetch(url, appendToVideoBuffer);

    indexVideoSegment++;

    if (indexVideoSegment > numberOfChunks) {
      videoSourceBuffer.removeEventListener('updateend', nextVideoSegment);
    }
  };

  // 3 - appendToVideoBuffer will get the chunk in param and add it to the buffer.
  const appendToVideoBuffer = (videoChunk: Iterable<number>) => {
    if (videoChunk) {
      videoSourceBuffer.appendBuffer(new Uint8Array(videoChunk));
    }
  };

  return (<>
    <video title="Video player" controls ref={playerRef} />
  </>)
}
```

### Step 4

Now that you should have a video playing you might realize that you don't have sound on it.
It is because we now need to do the same segment management with the audio.

<!-- TODO: Add gif a shit here we go again -->

You will then have a **videoSourceBuffer** and a **audioSourceBuffer** which will work just the same!

## 🎉 Congratulations!

You can now play a video like a pro! 🚀

## 🔎 What's next?

To be honest if you don't want to bother with all this chunk management, you can find libraries online that will handle it for you 😄

> If you want to have a look at some I would recommand to check:

Author: I think the right sentence is : 
> If you want to take a look at some, I would recommend **to check**

---
layout: post
title: "Crafting Your Own Streamer: A Hands-on Guide to JavaScript MSE"
description: In this article, I'm going to explain how to enhance your player using the Media Source Extension API.
author: [a_gaudard]
tags: [TVJS, smartTV, web, javascript, frontend, player, video]
color: rgb(251,87,66)
language: en
---

> Streaming is now everywhere and used by almost everyone, but do you know how it works behind the frame ?

Reviewer 2: ```suggestion
Streaming is now everywhere and used by almost everyone, but do you know how it works behind the frame?
```

Let say you want to create a player to watch your favorite TV Show.

First step will probably be to create a page one which you will add the famous video tag, give it a source and maybe display the controls so you can interact with it.

> You will end up with something like this :

Reviewer 2: ```suggestion
You will end up with something like this:
```

```html
<video src="" controls></video>
```

Your video will probably play well, but depending on, for example the video size or your connexion, you might have some performances issues.

That's where the Media Source Extension API comes to safe the day.

## 💡 Media Source Extension

_"MSE gives us finer-grained control over how much and how often content is fetched, and some control over memory usage details, such as when buffers are evicted."_ - [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/API/Media_Source_Extensions_API)

Here MDN tells us that MSE allows us to better control the flow of the stream.

> **This means new features are now available !**

Reviewer 2: ```suggestion
**This means new features are now available!**
```

> Going back to my previous use case, what if I want to play Dune in 4K ?

Reviewer 2: ```suggestion
Going back to my previous use case, what if I want to play Dune in 4K?
```

It is going to be a super heavy file and if I give it to my video tag, it is going to download the entire file before being ready to play the content.

> Do I want to wait to watch my movie ? Of course not ! I'm aleady in the sofa with my ice cream ready for it to start 🍦

Reviewer 2: ```suggestion
Do I want to wait to watch my movie? Of course not! I'm already in the sofa with my ice cream ready for it to start 🍦
```

> MSE helps us manipulate other video file format such as Dash (Dynamic Adaptive Streaming over HTTP) or HLS (HTTP Live Streaming). Those formats are a list of chunk of the video we want to play.
> So instead of having a massive video file we can now have a list of small part of the same video.

Reviewer 3: There is a little confusion on the keyword format and manifest. When we speak in my mind of format, i link it to the dash extension of segment and not the dash manifest MPD (Media Presentation Description).

```suggestion
MSE helps us manipulate other video file format such as Dash (Dynamic Adaptive Streaming over HTTP) or HLS (HTTP Live Streaming). Those formats split a video in a list of chunk and allow to play short section of the video at any timecode. Those chunk are referenced in manifest, MPD for dash and M3U8 for hls.
```

![Manifest file pointing to segments](/images/posts/2025-06-23-video-player-mse-introduction/manifest-1.png)

> That's not all ! Those list file (= manifest) can transport additional informations.

Reviewer 2: ```suggestion
That's not all! Those list file (= manifest) can transport additional informations.
```

> We can now have list of chunk based on specific settings such as the resolution, a recommanded bandwidth per list, a link to subtitles and many more !

Reviewer 2: ```suggestion
We can now have list of chunk based on specific settings such as the resolution, a recommanded bandwidth per list, a link to subtitles and many more!
```

Having multiple playlist based on the resoltion brings a new acronym (I now what you think another one). We can now talk about Adaptive Bitrate Streaming (= ABR).
This adds the feature of switching qualities based on the user bandwidth to our player 🤩

![Manifest file pointing to multiple playlist of segments](/images/posts/2025-06-23-video-player-mse-introduction/manifest-2.png)

Here is below an example of an HLS manifest where you can see the playlist links to video specific resolutions as well as links to the audio file and the subtitles.

![HLS manifest example](/images/posts/2025-06-23-video-player-mse-introduction/hls-manifest-example.png)

> ## 🤔 How to use it ?

Reviewer 2: ```suggestion
## 🤔 How to use it?
```

### Step 2

Now we want to add a **MediaSource** to handle our manifest.

Also, to better manipulate our video tag we use the useRef hook from React.

```js
import { useEffect, useRef } from "react";

export const Player = ({ source }: { source: string }) => {
  const playerRef = useRef<HTMLVideoElement>(null);

  // MediaSource instance
  let ms: MediaSource;

  useEffect(() => {
    ms = new MediaSource();

    // Link the MediaSource to the video tag
    playerRef.current!.src = window.URL.createObjectURL(ms);

    // 1 - Handle the sourceopen event
    ms.addEventListener('sourceopen', onMediaSourceOpen);
  });

  return (<>
    <video title="Video player" controls ref={playerRef} />
  </>)
}
```

### Step 3

Now we need to add a SourceBuffer to handle each segments of our video.

```js
import { useEffect, useRef } from "react";

export const Player = ({ source }: { source: string }) => {
  const playerRef = useRef<HTMLVideoElement>(null);

  let ms: MediaSource;

  // SourceBuffer instance
  let videoSourceBuffer: SourceBuffer;

  // 1 - Define the SourceBuffer
  const onMediaSourceOpen = () => {
    videoSourceBuffer = ms.addSourceBuffer('video/mp4; codecs="avc1.42c01e"');

>     // 3 - We add a new video chunk on each SourceBuffer update

Reviewer 2: There is no 2 😄

Author: Good catch ! 😄

    videoSourceBuffer.addEventListener('updateend', nextVideoSegment);

>     // 4 - We fetch the fist segment (init)

Reviewer 2: ```suggestion
    // 4 - We fetch the first segment (init)
```

    fetch(segmentInitVideo, appendToVideoBuffer);
  };

  useEffect(() => {
    ms = new MediaSource();

    playerRef.current!.src = window.URL.createObjectURL(ms);

    ms.addEventListener('sourceopen', onMediaSourceOpen);
  });

  return (<>
    <video title="Video player" controls ref={playerRef} />
  </>)
}
```

Here is the final version with the `nextVideoSegment` and `appendToVideoBuffer` methods.

```js
import { useEffect, useRef } from "react";
import { parseManifest, fetch } from './player.utils';

export const Player = ({ source }: { source: string }) => {
  const playerRef = useRef<HTMLVideoElement>(null);

  let ms: MediaSource;

  // SourceBuffer instance
  let videoSourceBuffer: SourceBuffer;

  //
  let indexVideoSegment = 1;

>   // You will need to parse the manifes to get the init segment as well as all the segments

Reviewer 2: ```suggestion
  // You will need to parse the manifest to get the init segment as well as all the segments
```

>   const { video: { segmentInit: segmentInitVideo, segments: segmentsVideo } } = parseManifest(source);

Reviewer 3: <!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**thought (non-blocking):** ‏
I understand that adding the code of the manifest parsing is not something we want on blog but i could be interesting to had a link in the ressource below with the project that use as exemple. Because in the case of someone wanting to reproduce the exemples given it will be difficult to do.  You could make an auto-promo to your github for the extra ressource.
Sadly the parsing is one of the things the more difficult to do to play video since there is no good lib to do it.

  // Here we just append 10 chunks
  const numberOfChunks = 10;

  // 1 - Define the SourceBuffer
  const onMediaSourceOpen = () => {
    videoSourceBuffer = ms.addSourceBuffer('video/mp4; codecs="avc1.42c01e"');

>     // 3 - We add a new video chunk on each SourceBuffer update

Reviewer 2: Still no 2

    videoSourceBuffer.addEventListener('updateend', nextVideoSegment);

    // 4 - We fetch the fist segment (init)
    fetch(segmentInitVideo, appendToVideoBuffer);
  };

  useEffect(() => {
    ms = new MediaSource();

    playerRef.current!.src = window.URL.createObjectURL(ms);

    ms.addEventListener('sourceopen', onMediaSourceOpen);
  });

  const nextVideoSegment = () => {
    var url = segmentsVideo.replace('$Number$', indexVideoSegment.toString());

    fetch(url, appendToVideoBuffer);

    indexVideoSegment++;

    if (indexVideoSegment > numberOfChunks) {
      videoSourceBuffer.removeEventListener('updateend', nextVideoSegment);
    }
  };

  const appendToVideoBuffer = (videoChunk: Iterable<number>) => {
    if (videoChunk) {
      videoSourceBuffer.appendBuffer(new Uint8Array(videoChunk));
    }
  };

  return (<>
    <video title="Video player" controls ref={playerRef} />
  </>)
}
```

### Step 4

Now that you should have a video playing you might realize that you don't have sound on it.
It is because we now need to do the same segment management with the audio.

<!-- TODO: Add gif a shit here we go again -->

> You will then have a **videoSourceBuffer** and a **audioSourceBuffer** which will work just the same !

Reviewer 2: ```suggestion
You will then have a **videoSourceBuffer** and a **audioSourceBuffer** which will work just the same!
```

## 🎉 Congratulations !

> You can now play a video like a pro ! 🚀

Reviewer 2: ```suggestion
You can now play a video like a pro! 🚀
```

> ## 🔎 What's next ?

Reviewer 2: ```suggestion
## 🔎 What's next?
```

To be honest if you don't want to bother with all this chunk management, you can find libraries online that will handle it for you 😄

> If you want to have a look at some I would recommand to check :

Reviewer 2: ```suggestion
If you want to have a look at some I would recommand to check:
```
